### PR TITLE
feat(voting-verifier): change votes to enums

### PR DESCRIPTION
**Pull request by eguajardo**
## Description

Currently votes are booleans. We need enums so we can represent more than 2 possible options.

https://axelarnetwork.atlassian.net/browse/AXE-2700

## Todos

- [ ] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

### CODE REVIEW

File: packages/axelar-wasm-std/src/voting.rs [hunk 1]
```diff
@@ -131,10 +133,49 @@ impl fmt::Display for PollId {
     }
 }
 
+#[cw_serde]
+#[derive(Eq, Hash, Ord, PartialOrd, EnumIter)]
+pub enum Vote {
+    Success,
+    Failure,
```
Comment (cjcobb23) [inline packages/axelar-wasm-std/src/voting.rs:140]:
```suggestion
    FailedOnChain, // the txn was included on chain, but failed to achieve the intended result
```

File: ampd/src/evm/verifier.rs [hunk 2]
```diff
@@ -83,21 +84,27 @@ pub fn verify_message(
...
-) -> bool {
+) -> Vote {
     match get_event(gateway_address, tx_receipt, msg.event_index) {
-        Some(event) => tx_receipt.transaction_hash == msg.tx_id && event == msg,
-        None => false,
+        Some(event) if tx_receipt.transaction_hash == msg.tx_id && event == msg => {
+            Vote::SucceededOnChain
+        }
+        Some(_) => Vote::NotFound,
+        None => Vote::NotFound,
```
Comment (fish-sammy) [inline ampd/src/evm/verifier.rs:93]:
```suggestion
        _ => Vote::NotFound,
```

File: ampd/src/evm/verifier.rs [hunk 3]
```diff
@@ -83,21 +84,27 @@ pub fn verify_message(
...
-) -> bool {
+) -> Vote {
     match get_event(gateway_address, tx_receipt, worker_set.event_index) {
-        Some(event) => tx_receipt.transaction_hash == worker_set.tx_id && event == worker_set,
-        None => false,
+        Some(event) if tx_receipt.transaction_hash == worker_set.tx_id && event == worker_set => {
+            Vote::SucceededOnChain
+        }
+        Some(_) => Vote::NotFound,
+        None => Vote::NotFound,
```
Comment (fish-sammy) [inline ampd/src/evm/verifier.rs:107]:
```suggestion
        _ => Vote::NotFound,
```

File: ampd/src/sui/verifier.rs [hunk 4]
```diff
@@ -55,19 +56,23 @@ pub fn verify_message(
...
+            if transaction_block.digest == message.tx_id
                 && event.type_ == call_contract_type(gateway_address)
-                && event == message
+                && event == message =>
+        {
+            Vote::SucceededOnChain
         }
-        None => false,
+        Some(_) => Vote::NotFound,
+        None => Vote::NotFound,
```
Comment (fish-sammy) [inline ampd/src/sui/verifier.rs:69]:
```suggestion
        _ => Vote::NotFound,
```

File: packages/axelar-wasm-std/src/voting.rs [hunk 5]
```diff
@@ -131,10 +134,49 @@ impl fmt::Display for PollId {
...
+
+impl Default for Tallies {
+    fn default() -> Self {
+        let mut tallies = BTreeMap::new();
+
+        Vote::iter().for_each(|vote| {
+            tallies.insert(vote, Uint256::zero());
+        });
+
+        Self(tallies)
```
Comment (fish-sammy) [inline packages/axelar-wasm-std/src/voting.rs:156]:
```suggestion
        Self(Vote::iter().map(|vote| (vote, Uint256::zero())).collect())
```

File: packages/axelar-wasm-std/src/voting.rs [hunk 6]
```diff
@@ -131,10 +134,49 @@ impl fmt::Display for PollId {
...
+    }
+}
+
+impl Tallies {
+    pub fn consensus(&self, quorum: Uint256) -> Option<Vote> {
+        self.0.iter().find_map(|(vote, tally)| {
+            if *tally >= quorum {
+                Some(vote.clone())
```
Comment (fish-sammy) [inline packages/axelar-wasm-std/src/voting.rs:164]: So it's possible that multiple types of votes exceed the quorum?

File: packages/axelar-wasm-std/src/voting.rs [hunk 7]
```diff
@@ -131,10 +134,49 @@ impl fmt::Display for PollId {
...
+            } else {
+                None
+            }
+        })
+    }
+
+    pub fn tally(&mut self, vote: &Vote, weigth: &Uint256) {
```
Comment (fish-sammy) [inline packages/axelar-wasm-std/src/voting.rs:171]:
```suggestion
    pub fn tally(&mut self, vote: &Vote, weight: &Uint256) {
```

File: packages/axelar-wasm-std/src/voting.rs [hunk 8]
```diff
@@ -207,14 +249,26 @@ impl WeightedPoll {
...
             .filter_map(|(address, participation)| {
-                participation.vote.as_ref().and_then(|vote| {
-                    if *vote == results {
+                participation.vote.as_ref().and_then(|votes| {
+                    let voted_consensus = votes.iter().zip(results.iter()).all(|(vote, result)| {
+                        if let Some(result) = result {
+                            vote == result
+                        } else {
+                            true // if there was no consensus, we don't care about the vote
+                        }
```
Comment (fish-sammy) [inline packages/axelar-wasm-std/src/voting.rs:268]:
```suggestion
                        result.is_none() || Some(vote) == result.as_ref()
```

**Review by fish-sammy — CHANGES_REQUESTED** [review]

File: packages/axelar-wasm-std/src/voting.rs (hunk 6 above)
Comment (eguajardo) [reply packages/axelar-wasm-std/src/voting.rs]: Task to enforce majority in thresholds: https://axelarnetwork.atlassian.net/browse/AXE-2832

File: packages/axelar-wasm-std/src/voting.rs [hunk 9]
```diff
@@ -131,10 +135,62 @@ impl fmt::Display for PollId {
...
+        match self {
+            Vote::SucceededOnChain => write!(f, "SucceededOnChain"),
+            Vote::FailedOnChain => write!(f, "FailedOnChain"),
+            Vote::NotFound => write!(f, "NotFound"),
+        }
+    }
+}
+
+// Deserialization of enums as map keys is not supported by serde-json-wasm, we use String instead
+type VoteString = String;
```
Comment (fish-sammy) [inline packages/axelar-wasm-std/src/voting.rs:157]: I don't see the point of creating this type alias.

File: packages/axelar-wasm-std/src/voting.rs [hunk 10]
```diff
@@ -131,10 +135,62 @@ impl fmt::Display for PollId {
...
+                .collect(),
+        )
+    }
+}
+
+impl Tallies {
+    pub fn consensus(&self, quorum: Uint256) -> Option<Vote> {
+        self.0.iter().find_map(|(vote, tally)| {
+            if *tally >= quorum {
+                Some(Vote::from_str(vote).expect("can't parse vote string back to enum"))
```
Comment (fish-sammy) [inline packages/axelar-wasm-std/src/voting.rs:175]:
```suggestion
                Some(vote.parse().expect("can't parse vote string back to enum"))
```

**Review by fish-sammy — APPROVED** [review]

**Pull request merged**